tests/math/test_density_matrices.py

```diff
@@ -210,3 +210,160 @@ def test_density_matrix_tf_jit(self):
         )
         density_matrix = density_matrix(state_vector)
         assert np.allclose(density_matrix, [[1, 0], [0, 0]])
+
+
+# fmt: off
+density_matrices = [
+    (mat_00, (mat_0, mat_0)),
+    (mat_01, (mat_0, mat_1)),
+    (mat_10, (mat_1, mat_0)),
+    (mat_11, (mat_1, mat_1)),
+    (onp.array(mat_00), (mat_0, mat_0)),
+    (onp.array(mat_01), (mat_0, mat_1)),
+    (onp.array(mat_10), (mat_1, mat_0)),
+    (onp.array(mat_11), (mat_1, mat_1)),
+    (np.array(mat_00), (mat_0, mat_0)),
+    (np.array(mat_01), (mat_0, mat_1)),
+    (np.array(mat_10), (mat_1, mat_0)),
+    (np.array(mat_11), (mat_1, mat_1)),
+    (jnp.array(mat_00), (mat_0, mat_0)),
+    (jnp.array(mat_01), (mat_0, mat_1)),
+    (jnp.array(mat_10), (mat_1, mat_0)),
+    (jnp.array(mat_11), (mat_1, mat_1)),
+    (torch.tensor(mat_00), (mat_0, mat_0)),
+    (torch.tensor(mat_01), (mat_0, mat_1)),
+    (torch.tensor(mat_10), (mat_1, mat_0)),
+    (torch.tensor(mat_11), (mat_1, mat_1)),
+    (tf.Variable(mat_00), (mat_0, mat_0)),
+    (tf.Variable(mat_01), (mat_0, mat_1)),
+    (tf.Variable(mat_10), (mat_1, mat_0)),
+    (tf.Variable(mat_11), (mat_1, mat_1)),
+    (tf.constant(mat_00), (mat_0, mat_0)),
+    (tf.constant(mat_01), (mat_0, mat_1)),
+    (tf.constant(mat_10), (mat_1, mat_0)),
+    (tf.constant(mat_11), (mat_1, mat_1)),
+    (mat_00_10, (mat_0_1, mat_0)),
+    (mat_01_11, (mat_0_1, mat_1)),
+]
+
+# fmt: on
+
+
+class TestDensityMatrixFromMatrix:
```

Comment:
As stated in the other PRs, it would be good to test that the dtype of the returned matrix matches the user-provided dtype.